src/slice.rs

```diff
@@ -629,3 +630,135 @@ macro_rules! s(
+pub trait MultiSlice<'a, A, D>
+where
+    A: 'a,
+    D: Dimension,
+{
+    /// The type of the slices created by `.multi_slice_move()`.
+    type Output;
+
+    /// Slice the raw view into multiple raw views, and dereference them.
+    ///
+    /// **Panics** if performing any individual slice panics or if the slices
+    /// are not disjoint (i.e. if they intersect).
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output;
+}
+
+impl<'a, A, D> MultiSlice<'a, A, D> for ()
+where
+    A: 'a,
+    D: Dimension,
+{
+    type Output = ();
+
+    fn multi_slice_move(&self, _view: ArrayViewMut<'a, A, D>) -> Self::Output {}
+}
+
+impl<'a, A, D, Do0> MultiSlice<'a, A, D> for (SliceInfo<D::SliceArg, Do0>,)
+where
+    A: 'a,
+    D: Dimension,
+    D::SliceArg: Sized,
+    Do0: Dimension,
+{
+    type Output = (ArrayViewMut<'a, A, Do0>,);
+
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
+        (view.slice_move(&self.0),)
+    }
+}
+
+impl<'a, A, D, Do0> MultiSlice<'a, A, D> for (&SliceInfo<D::SliceArg, Do0>,)
+where
+    A: 'a,
+    D: Dimension,
+    Do0: Dimension,
+{
+    type Output = (ArrayViewMut<'a, A, Do0>,);
+
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
+        (view.slice_move(self.0),)
+    }
+}
+
+macro_rules! impl_multislice_tuple {
+    ($($Do:ident,)*) => {
+        impl<'a, A, D, $($Do,)*> MultiSlice<'a, A, D> for ($(SliceInfo<D::SliceArg, $Do>,)*)
+        where
+            A: 'a,
+            D: Dimension,
+            D::SliceArg: Sized,
+            $($Do: Dimension,)*
+        {
+            type Output = ($(ArrayViewMut<'a, A, $Do>,)*);
+
+            fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
+                #[allow(non_snake_case)]
+                let ($($Do,)*) = self;
+
+                let shape = view.raw_dim();
+                assert!(!impl_multislice_tuple!(@intersects_self &shape, ($(&$Do,)*)));
+
+                let raw_view = view.into_raw_view_mut();
+                unsafe {
+                    ($(raw_view.clone().slice_move(&$Do).deref_into_view_mut(),)*)
+                }
+            }
+        }
+
+        impl<'a, A, D, $($Do,)*> MultiSlice<'a, A, D> for ($(&SliceInfo<D::SliceArg, $Do>,)*)
+        where
+            A: 'a,
+            D: Dimension,
+            $($Do: Dimension,)*
+        {
+            type Output = ($(ArrayViewMut<'a, A, $Do>,)*);
+
+            fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
+                #[allow(non_snake_case)]
+                let ($($Do,)*) = self;
+
+                let shape = view.raw_dim();
+                assert!(!impl_multislice_tuple!(@intersects_self &shape, ($($Do,)*)));
+
+                let raw_view = view.into_raw_view_mut();
+                unsafe {
+                    ($(raw_view.clone().slice_move($Do).deref_into_view_mut(),)*)
```

Comment:
So rather less important, but if we peel off one `$Do` from the end of the repeat, we can save one clone() call. And as it happens there are always at least 2 $Do.

Reply:
This is possible, but unfortunately the obvious approach (using a matcher like `($($but_last:ident,)* $last:ident,)`) doesn't work due to rust-lang/rust#24827/rust-lang/rust#42838. The best solution I see is to build the expression in macro parameters, like this:

```rust
// ...
                let raw_view = view.into_raw_view_mut();
                unsafe {
                    impl_multislice_tuple!(@build_slice_expr raw_view, ($($Do,)*), ())
                }
            }
        }
    };
    (@build_slice_expr $raw_view:ident, ($head:expr,), ($($slice:expr,)*)) => {
        ($($slice,)* $raw_view.slice_move($head).deref_into_view_mut(),)
    };
    (@build_slice_expr $raw_view:ident, ($head:expr, $($tail:expr,)*), ($($slice:expr,)*)) => {
        impl_multislice_tuple!(
            @build_slice_expr $raw_view, ($($tail,)*),
            (
                $($slice,)*
                $raw_view.clone().slice_move($head).deref_into_view_mut(),
            )
        )
    };
// ...
```

but I'm not sure it's worth the additional complexity, especially since that complexity is in an `unsafe` block. Maybe there's a better way to do it?

Reply:
I found a better way. (Just separate into "but last" and "last" in the original macro call; don't make the macro try to do it.) I've updated the PR.